feat: Windows Support

JeremyEastham opened the pull request:
## Summary of Change
Trying to contribute to this project has been a challenge to set up on Windows. The purpose of this PR is to streamline the project setup process for other developers that use Windows on their development machine. I have fixed a number of Windows-specific bugs and added extensive Windows development setup instructions to `CONTRIBUTING.md` (for both Git Bash and WSL2).

Before starting work on this PR, **only 87% of tests were passing when run on Windows** via Git Bash. I chose to use Git Bash because it uses the native Windows version of Java at runtime, while still being able to execute the bash scripts in `supertokens-root`. It was important to me that the native Windows version of Java was used so that I could make sure that all core functionality remained intact when the core was run on a Windows machine. Many of the tests were failing due to bugs in `Utils.java`, which did not use `System.lineSeparator()` correctly when mocking config values. Additionally, a few tests compared exception message strings that were not identical cross-platform.

## Related Issues
(no related issues)

## Test Plan
I have removed a portion of `WebserverTest.differentHostNameTest()`. This test was failing because the core was not visible to any of my local IP addresses when hosted on `localhost` due to my firewall settings. Thus, it was not an issue with the code, but instead, an issue with the test. This may be related to how Windows handles `localhost` addresses, but documentation is sparse on this issue.

The purpose of this PR is not to add functionality, so no tests have been added.

## Documentation Changes
(no changes needed)

## Checklist for important updates
- [X] Changelog has been updated
    - [X] If there are any db schema changes, mention those changes clearly
- [X] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [X] `pluginInterfaceSupported.json` file has been updated (if needed)
- [X] Changes to the version if needed
   - In `build.gradle`
- [X] Had installed and ran the pre-commit hook
- [X] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [X] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Figure out why "Running the Core (Manually)" instructions are not working
    - [ ] Rewrite instructions in `CONTRIBUTING.md` if needed
- [ ] Merge supertokens/supertokens-root#8
    - Instructions added to `CONTRIBUTING.md` are dependent on these run configurations existing

rishabhpoddar commented:
@JeremyEastham please also make sure to run the tests with your modifications via github actions so that the failing PR check passes.

JeremyEastham commented:
Tests are failing now because I renamed `./startTestingEnv` to `./startTestEnv` in supertokens/supertokens-root#9. Once that PR is merged, tests *should* pass. The problem is that I changed the call in *this* PR to `./startTestEnv`, which does not exist in supertokens-root master yet.

rishabhpoddar commented on src/test/java/io/supertokens/test/Utils.java:
```diff
@@ -31,22 +32,21 @@
 
 import java.io.*;
 import java.nio.charset.StandardCharsets;
```
Changes to this file should also be reflected in the plugin repos for the core.

JeremyEastham replied:
Although I changed the implementations of these methods to make them more readable, the method signatures did not change. No changes should be needed in the plugin repos.

JeremyEastham commented on src/test/java/io/supertokens/test/AuthRecipeTest.java:
```diff
@@ -205,8 +205,11 @@ public void paginationTest() throws Exception {
         }
 
         UserInfo user1 = EmailPassword.signUp(process.getProcess(), "test0@example.com", "password0");
+        Thread.sleep(1); // Sleep to remove race condition
```
One of the times that I ran this test, the test failed because when the users were sorted by creation time, they were not sorted in the correct order. This is due to Java's implementation of `System.currentTimeMillis()` that does not necessarily return a unique timestamp on each call.

See this excerpt from the Java documentation:
> `public static long currentTimeMillis()`
> 
> Returns the current time in milliseconds. Note that while the unit of time of the return value is a millisecond, the granularity of the value depends on the underlying operating system and may be larger. For example, many operating systems measure time in units of tens of milliseconds.

For Example:
```java
for(int i = 0; i < 50; i++)
    System.out.println(System.currentTimeMillis());
```
When I ran this code on my Windows machine, it printed out `1650942035022` many times, followed by `1650942035026` many more times.

However, if I run this second example:
```java
for(int i = 0; i < 50; i++) {
    System.out.println(System.currentTimeMillis());
    Thread.sleep(1);
}
```
The above prints out 50 unique timestamps, 1-3 milliseconds apart. This is due to Java's implementation of `Thread.sleep()`, which is guaranteed to sleep *at least* the number of milliseconds that it is given, dependent on the resolution of the clock of the underlying operating system.

JeremyEastham commented on src/test/java/io/supertokens/test/WebserverTest.java:
```diff
@@ -597,26 +595,12 @@ public void differentHostNameTest() throws InterruptedException, IOException, Ht
         Utils.setValueInConfig("host", "\"localhost\"");
         hello("localhost", "3567");
         hello("127.0.0.1", "3567");
-        try {
-            hello(inetAddress.getHostAddress(), "3567");
-            if (!inetAddress.getHostAddress().equals("127.0.0.1")) {
-                fail();
-            }
-        } catch (ConnectException ignored) {
-        }
 
         Utils.reset();
 
         Utils.setValueInConfig("host", "\"127.0.0.1\"");
         hello("localhost", "3567");
         hello("127.0.0.1", "3567");
-        try {
-            hello(inetAddress.getHostAddress(), "3567");
-            if (!inetAddress.getHostAddress().equals("127.0.0.1")) {
-                fail();
-            }
-        } catch (ConnectException ignored) {
-        }
```
This test was failing because the core was not visible to any of my local IP addresses when hosted on localhost due to my firewall settings. This may also be related to how Windows handles localhost addresses, but documentation is sparse on this issue.

rishabhpoddar replied:
I see. I would require these tests to be there.. Maybe we can change the code here to not run if the OS is windows?

JeremyEastham replied:
I thought about that, but wouldn't the situation be similar on other operating systems as well? If settings are strict (as they probably should be), access to localhost servers would not be visible to LAN addresses.

rishabhpoddar replied:
It has more or less worked with all linux based systems, mac, CICD, GitHub action and everyone in our dev team.. so i don't think it's too much of a problem. I just want this to also run in CICD / GH action pipeline. It would be OK to have them skipped in local dev env.

rishabhpoddar commented on src/test/java/io/supertokens/test/Utils.java:
```diff
@@ -138,46 +137,26 @@ public static void reset() {
-                writer.write(modifiedFileContent);
-            }
-        }
+        String find = "\r?\n#?\\s*" + Pattern.quote(key) + ":.*\r?\n";
+        String replace = newLine + key + ": " + value + newLine;
```
Do we need to change the regex expression? The older one worked just fine. Did the older one cause issues in windows?

JeremyEastham replied:
Yes, this change is necessary.

```
Old Regex: "\n((#\\s)?)" + key                + "(:|((:\\s).+))\n"
New Regex: "\r?\n#?\\s*" + Pattern.quote(key) + ":.*\r?\n"
```

Issues with the Old Regex:
- It used `\n` instead of `\r?\n`, which did not match Windows line endings (`\r\n`)
- It only matched comments with a single whitespace character after it
- It did not match non-commented keys with leading whitespace
- It was also much harder to read due to unnecessary group nesting and the unnecessary `|` operator
- It did not quote the key, so any regex characters in the key would make it fail

When reviewing this, I actually noticed one issue with the New Regex. It did not match commented lines with leading whitespace. I have committed this change.

rishabhpoddar replied:
Ok. So this change would be required in the plugins repos as well. Since they too have a similar function in there.

JeremyEastham replied:
Okay, I see. I was thinking that the plugin repos called `io.supertokens.test.Utils.commentConfigValue()` and `io.supertokens.test.Utils.setValueInConfig()` directly.

I was also thinking about renaming `setValueInConfig()` to `setConfigValue()` to match `commentConfigValue()`. If I rename all references in `supertokens-core`, will any change be needed in other repositories?

rishabhpoddar replied:
You should change them in the plugin repos too, just to be consistent.

rishabhpoddar commented on CONTRIBUTING.md:
```diff
@@ -3,118 +3,232 @@
+6. Click on "Run workflow"
 
-## Running the core manually
+## Running the Core (Manually)
```
```suggestion
## Running the Core (Using Java)

```

JeremyEastham replied:
I have clarified this in this document. It now reads `With Console`.

rishabhpoddar commented on CONTRIBUTING.md:
```diff
@@ -3,118 +3,232 @@
 
-## Running the core manually
+## Running the Core (Manually)
 1. Run `startTestingEnv --wait` in a terminal, and keep it running
```
```suggestion
1. Run `startTestEnv --wait` in a terminal, and keep it running

```

JeremyEastham replied:
This part of `CONTRIBUTING.md` has been replaced by instructions to run `runCore`.

rishabhpoddar commented on CONTRIBUTING.md:
```diff
@@ -3,118 +3,232 @@
 1. Run `startTestingEnv --wait` in a terminal, and keep it running
 2. Then open `supertokens-root` in another terminal and run `cp ./temp/config.yaml .`
-3. Then run `java -classpath "./core/*:./plugin-interface/*" io.supertokens.Main ./ DEV`. This will start the core to listen on `http://localhost:3567`
+3. Then run `java -classpath "./supertokens-core/*:./supertokens-plugin-interface/*" io.supertokens.Main ./ DEV`. This will start the core to listen on `http://localhost:3567`
```
Are you sure this change is right?

JeremyEastham replied:
This is incorrect. I wrote this before I wrote the `runCore` script. I'm going to replace these instructions with instructions to run `runCore`.

rishabhpoddar commented on CONTRIBUTING.md:
```diff
@@ -3,118 +3,232 @@
 2. Then open `supertokens-root` in another terminal and run `cp ./temp/config.yaml .`
-3. Then run `java -classpath "./core/*:./plugin-interface/*" io.supertokens.Main ./ DEV`. This will start the core to listen on `http://localhost:3567`
+3. Then run `java -classpath "./supertokens-core/*:./supertokens-plugin-interface/*" io.supertokens.Main ./ DEV`. This will start the core to listen on `http://localhost:3567`
 
```
We should also have instrs on running the core via the `./runCore` script.

JeremyEastham replied:
See above.
